scipy/optimize/_slsqp_py.py

```diff
@@ -529,15 +525,16 @@ def cjac(x, *args):
         kkt_multiplier[_t] = kkt
 
     # Get KKT multipliers for lower and upper bounds.
-    kkt = np.zeros((n, 2))
-    pos_kkt_bnds = _kkt_bnds > 0.
-    # Lower bound multipliers are positive
-    active_bnds = np.isfinite(xl) & pos_kkt_bnds
-    kkt[active_bnds, 0] = _kkt_bnds[active_bnds]
-    # Upper bound multipliers are originally negative, make positive to be
-    # consistent with sign of multipliers for inequality constraints
-    active_bnds = np.isfinite(xu) & ~pos_kkt_bnds
-    kkt[active_bnds, 1] = -_kkt_bnds[active_bnds]
+    kkt = zeros((n, 2))
+    if bounds is not None and len(bounds) > 0:
+        pos_kkt_bnds = _kkt_bnds > 0.
+        # Lower bound multipliers are positive
+        active_bnds = ~infbnd[:, 0] & pos_kkt_bnds
+        kkt[active_bnds, 0] = _kkt_bnds[active_bnds]
+        # Upper bound multipliers are originally negative, make positive to be
+        # consistent with sign of multipliers for inequality constraints
+        active_bnds = ~infbnd[:, 1] & ~pos_kkt_bnds
+        kkt[active_bnds, 1] = np.abs(_kkt_bnds[active_bnds])
```

Comment:
What motivated this change?

Reply:
Sometimes `pos_kkt_bnds` would be `True` when `_kkt_bnds == 0`, and then these got changed to `-0`. The `abs` takes care of this in one line.

Reply:
My concern is that KKT multipliers can have signs. For example, in the example you added, the sign tells you which direction the objective function will change when you adjust the constraints. Without reading carefully, it looks like this will interfere with that, no?

Reply:
Based on your comment [https://github.com/scipy/scipy/pull/15641#issuecomment-1189881932](https://github.com/scipy/scipy/pull/15641#issuecomment-1189881932), which I verified from a few sources, multipliers for _equality_ constraints can be either positive or negative, whereas multipliers for _inequality_ constraints must be positive (if the KKT conditions hold).

Since bound constraints of the form $a \leq x \leq b$ can be viewed as a pair of inequality constraints, $x - a \geq 0$ and $b - x \geq 0$, it follows that the KKT multipliers associated to bound constraints should always be positive, whether or not the bound is a lower bound or upper bound.

On the other hand, the fortran code which finds the KKT multipliers for the bound constraints uses a working space of size `n`, where `n = size(x)`. This works because only one bound will be active at a time. It marks active lower bounds with positive multipliers and active upper bounds with negative multipliers.

This presents us with two options:
1. Return `kkt['bounds']` as a single array, keeping the sign convention used by the fortran code.
2. Return separate KKT multipliers for lower and upper bounds, which I've done. To make these multipliers consistent with the conventions for inequality constraints as described above, make them all non-negative.

As you can see I went for the second choice, but I'm open to discussion!

In terms of interpreting the KKT multipliers, suppose the inequality constraint $g(x) \geq 0$ is active at the solution $x^*$. If we tighten the constraint by $\epsilon > 0$, i.e. $g(x) \geq \epsilon$ or $g(x) - \epsilon \geq 0$, we expect the objective $f(x^*)$ to increase by $\epsilon \mu$. This is what your example verifies in practice.

Looking at lower bound constraints, $g(x) = x - a \geq 0$, if we tighten the constraint to $x \geq a + \epsilon$ then we expect $f(x^*)$ to increase so the lower bound KKT multipliers must be positive. Similarly, if we write upper bound constraints as $g(x) = b - x \geq 0$, tightening the constraint transforms this into $b - \epsilon \geq x$. This again increases $f(x^*)$ so these KKT multipliers should also be positive.

To your point, when we write $a \leq x \leq b$ we _could_ think of moving both bounds in the same direction by $\epsilon$, thereby _loosening_ the upper bound to $x \leq b + \epsilon$. This would decrease $f(x^*)$, which is probably why the fortran code makes these multipliers negative. I think both interpretations are valid, but the former (all positive multipliers) is more consistent.

Reply:
Thanks. I don't think I thought about it at the time, but here is what I understand now: if an inequality constraint is active, "tightening" it can _only_ make the optimal value of the objective function worse. (After all, if moving the solution in that direction were to improve the objective function, the constraint would not be active!) So we can assign whatever sign to the inequality constraint multipliers that we wish to as long as we keep the above in mind when we interpret it.

> As you can see I went for the second choice, but I'm open to discussion!

The second choice is what I would have done. I'm happy with that, and the absolute value sign is fine. I just got concerned because previously, there was an absolute value sign on the multipliers corresponding with an _equality_ constraint, and that was not right.